### DISCUSSION TITLE
Persist cache volumes into a container e.g. withCache

### DISCUSSION THREAD

**Discussion by RonanQuigley** [dagger/dagger#12470] [Help]
### What are you trying to do?

Let's say I have a node app. 

At present, I can use `withMountedCache` to reduce the npm install times i.e. it does not require us to re-fetch the dependencies over the internet. 

```js


```json
{
  "name": "test",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "express": "^4.18.2"
  }
}
```

Node apps require the node modules folder to be shipped with the image to run code that relies on third party node modules:

```
import * as dagger from '@dagger.io/dagger';

dagger.connect(
  async (client) => {
    const cacheVolume = client.cacheVolume('test');

    await client
      .container()
      .from('node:16-slim')
      .withEntrypoint(['/bin/sh', '-c'])
      .withWorkdir('/test')
      .withDirectory('.', client.host().directory('.'), {
        exclude: ['node_modules'],
      })
      // can't use withMountedCache for caching as it will never be in the published image
      // which is needed for the app to run
      // we need a withCache equivalent
      .withMountedCache('/test/node_modules', cacheVolume)
      .withExec('npm i')
      .withExec('ls -lat node_modules')
      .publish('image/test:latest');
  },
  {
    LogOutput: process.stdout,
    Workdir: '.',
  }
);
```

The published image won't work as `withMountedCache` is a mount. Mounts in dagger i.e. withMountedDirectory, withMountedFile etc. do not persist their contents into the container. 

@kpenfound @jpadams 

### Why is this important to you?

I want to be robust with caching, but I can't ship Node images that use node_modules without the node_modules folder as they'll crash. 

### How are you currently working around this?

Using withDirectory.

However, this is _not_ a good solution. It's not obvious in my repro, but at scale e.g. 100Mbs worth of node modules, re-fetching all dependencies and doing a full reinstall is slow. 

Another option is a `withExec` + copying the results of the install into another folder within the container. This is also a bad solution as recursively copying node modules is :snail:

**Comment by marcosnils** [comment]
Hey Ronan! 

As discussed in Slack, the best way to accomplish this is not to send the `node_modules` folder over the build context but to use a `WithMountedCache` to let the package manager populate the `node_modules` folder correctly. Here's the example I've shared in Discord

```javascript
import { connect } from "@dagger.io/dagger"

// initialize Dagger client
connect(
  async (client) => {

    const assets = client.host().directory(".", { exclude: ["node_modules", "out"] })

    const node = client.container().from("node:16")
      .withExec(["node", "-v"])
      .withMountedCache("/usr/local/share/.cache/yarn", client.cacheVolume("yarn_cache"))
      .withDirectory("/app", assets)
      .withWorkdir("/app")
      .withExec(["yarn", "install"])


    // execute
    await node.directory("/app").export("./out")
  },
  { LogOutput: process.stderr }
)
```

as you can see, I'm using `withMountedCache` and then I let `yarn install` create the node_modules folder by linking the yarn cache assets which will be present upon exporting the application directory. Would this work in your use case?
